### DISCUSSION TITLE
Problem with building zephyr/samples/subsys/mgmt/mcumgr/smp_svr  using atsame70

### DISCUSSION THREAD

**Discussion by CEmir10** [zephyrproject-rtos/zephyr#46144] [General] [answered]
I was able to build and flash same example on nrf52840dk_nrf8240, i was also able to play with mcumgr where i upload new image. My task is to do it wit atsame70, but when i try(west build -b psn_jlink -- -DOVERLAY_CONFIG='overlay-serial.conf;overlay-fs.conf;overlay-shell-mgmt.con)  sometimes I get this error(when serial transport is enabled): In file included from /home/crem/zephyrproject/zephyr/include/arch/arm/aarch32/arch.h:20,
                 from /home/crem/zephyrproject/zephyr/include/arch/cpu.h:19,
                 from /home/crem/zephyrproject/zephyr/include/kernel_includes.h:33,
                 from /home/crem/zephyrproject/zephyr/include/kernel.h:17,
                 from /home/crem/zephyrproject/zephyr/include/zephyr.h:18,
                 from /home/crem/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/src/main.c:8:
/home/crem/zephyrproject/zephyr/include/devicetree/fixed-partitions.h:53:9: error: 'DT_COMPAT_fixed_partitions_LABEL_storage_PARTITION_ID' undeclared here (not in a function)
   53 |  DT_CAT(DT_COMPAT_fixed_partitions_LABEL_, label)
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/crem/zephyrproject/zephyr/include/devicetree.h:3022:24: note: in definition of macro 'DT_CAT'
 3022 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/crem/zephyrproject/zephyr/include/storage/flash_map.h:266:2: note: in expansion of macro 'DT_FIXED_PARTITION_ID'
  266 |  DT_FIXED_PARTITION_ID(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
      |  ^~~~~~~~~~~~~~~~~~~~~
/home/crem/zephyrproject/zephyr/include/devicetree/fixed-partitions.h:53:2: note: in expansion of macro 'DT_CAT'
   53 |  DT_CAT(DT_COMPAT_fixed_partitions_LABEL_, label)
      |  ^~~~~~
/home/crem/zephyrproject/zephyr/include/storage/flash_map.h:266:24: note: in expansion of macro 'DT_NODE_BY_FIXED_PARTITION_LABEL'
  266 |  DT_FIXED_PARTITION_ID(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/crem/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/src/main.c:58:25: note: in expansion of macro 'FLASH_AREA_ID'
   58 |  .storage_dev = (void *)FLASH_AREA_ID(storage),
      |                         ^~~~~~~~~~~~~
[24/229] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/rb.c.obj
ninja: build stopped: subcommand failed.
but sometimes this one: In file included from /home/crem/zephyrproject/zephyr/include/toolchain/gcc.h:72,
                 from /home/crem/zephyrproject/zephyr/include/toolchain.h:50,
                 from /home/crem/zephyrproject/zephyr/lib/libc/minimal/include/string.h:13,
                 from /home/crem/zephyrproject/zephyr/drivers/console/uart_mcumgr.c:12:
/home/crem/zephyrproject/zephyr/drivers/console/uart_mcumgr.c: In function 'uart_mcumgr_register':
/home/crem/zephyrproject/zephyr/include/device.h:96:39: error: '__device_dts_ord_DT_CHOSEN_zephyr_uart_mcumgr_ORD' undeclared (first use in this function)
   96 | #define DEVICE_NAME_GET(name) _CONCAT(__device_, name)
      |                                       ^~~~~~~~~
/home/crem/zephyrproject/zephyr/include/device.h:291:37: note: in expansion of macro 'DEVICE_NAME_GET'
  291 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_NAME(node_id))
      |                                     ^~~~~~~~~~~~~~~
/home/crem/zephyrproject/zephyr/include/device.h:311:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  311 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/home/crem/zephyrproject/zephyr/drivers/console/uart_mcumgr.c:249:20: note: in expansion of macro 'DEVICE_DT_GET'
  249 |  uart_mcumgr_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_uart_mcumgr));
      |                    ^~~~~~~~~~~~~
/home/crem/zephyrproject/zephyr/include/device.h:96:39: note: each undeclared identifier is reported only once for each function it appears in
   96 | #define DEVICE_NAME_GET(name) _CONCAT(__device_, name)
      |                                       ^~~~~~~~~
/home/crem/zephyrproject/zephyr/include/device.h:291:37: note: in expansion of macro 'DEVICE_NAME_GET'
  291 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_NAME(node_id))
      |                                     ^~~~~~~~~~~~~~~
/home/crem/zephyrproject/zephyr/include/device.h:311:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  311 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/home/crem/zephyrproject/zephyr/drivers/console/uart_mcumgr.c:249:20: note: in expansion of macro 'DEVICE_DT_GET'
  249 |  uart_mcumgr_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_uart_mcumgr));
      |                    ^~~~~~~~~~~~~
ninja: build stopped: subcommand failed.

**Comment by mbolivar-nordic** [comment, marked as answer]
@CEmir10 you seem to be using a custom board, and based on the error output, it looks like the board's devicetree is missing the appropriate nodes to configure flash partitions. I'm converting this to a discussion since it does not look like a bug in upstream zephyr itself.

**Reply by CEmir10** [reply]
Right, this is a custom board.

**Reply by CEmir10** [reply]
do you know what i should modify, can you try to build it for sam_e70_xplained?

**Reply by mbolivar-nordic** [reply]
I don't have any experience with that platform, sorry. You may want to ask @nandojve or on Discord.

**Reply by CEmir10** [reply]
Should i send him question directly or he can answer even here?

**Reply by nandojve** [reply]
Hi @CEmir10 ,

There is no official support to MCUboot and MCUMGR for any SAM series. I took a small look at sample and I noted this: https://github.com/zephyrproject-rtos/zephyr/blob/c62ee3be4e4f9c268a009bced3b4d969d7ccbe46/samples/subsys/mgmt/mcumgr/smp_svr/usb.overlay#L7-L11

Which means that you must declare that node to bind with your serial port. The node is documented here https://docs.zephyrproject.org/latest/build/dts/api/api.html?highlight=zephyr%20uart%20mcumgr#id2

Maybe `zephyr,uart-mcumgr` node requirement could be added in the `Prerequisites` topic of the sample to make this cleaner.

**Comment by 0xff07** [comment]
I have the same issue compiling this example even for some default targets. For example, if I choose `nucleo_f411re` as target:

```
west build -p -b nucleo_f411re -- -DCONF_FILE="prj.conf overlay-bt.conf"
```

then I got following errors:

```
[50/266] Building C object CMakeFiles/app.dir/src/main.c.obj
FAILED: CMakeFiles/app.dir/src/main.c.obj 
ccache /Users/0xff19/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DCORE_CM4 -DHSE_VALUE=8000000 -DKERNEL -DLFS_CONFIG=zephyr_lfs_config.h -DSTM32F411xE -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -I/Users/0xff19/zephyrproject/zephyr/include -I/Users/0xff19/Workspace/zephyr_apps/smp_svr/build/zephyr/include/generated -I/Users/0xff19/zephyrproject/zephyr/soc/arm/st_stm32/stm32f4 -I/Users/0xff19/zephyrproject/zephyr/lib/util/fnmatch/. -I/Users/0xff19/zephyrproject/zephyr/drivers -I/Users/0xff19/zephyrproject/zephyr/soc/arm/st_stm32/common -I/Users/0xff19/zephyrproject/zephyr/subsys/bluetooth -I/Users/0xff19/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include -I/Users/0xff19/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/soc -I/Users/0xff19/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include -I/Users/0xff19/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include/Legacy -I/Users/0xff19/zephyrproject/modules/hal/stm32/stm32cube/common_ll/include -I/Users/0xff19/zephyrproject/modules/crypto/tinycrypt/lib/include -I/Users/0xff19/zephyrproject/modules/lib/zcbor/include -I/Users/0xff19/zephyrproject/modules/fs/littlefs -I/Users/0xff19/zephyrproject/zephyr/modules/littlefs -I/Users/0xff19/zephyrproject/zephyr/subsys/mgmt/mcumgr/lib/mgmt/include -I/Users/0xff19/zephyrproject/zephyr/subsys/mgmt/mcumgr/lib/util/include -I/Users/0xff19/zephyrproject/zephyr/subsys/mgmt/mcumgr/lib/smp/include -I/Users/0xff19/zephyrproject/zephyr/subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/include -I/Users/0xff19/zephyrproject/zephyr/subsys/mgmt/mcumgr/lib/cmd/img_mgmt/include -I/Users/0xff19/zephyrproject/zephyr/subsys/mgmt/mcumgr/lib/cmd/os_mgmt/include -I/Users/0xff19/zephyrproject/zephyr/subsys/mgmt/mcumgr/lib/cmd/stat_mgmt/include -isystem /Users/0xff19/zephyrproject/zephyr/lib/libc/minimal/include -isystem /Users/0xff19/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include -isystem /Users/0xff19/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include-fixed -Os -imacros /Users/0xff19/Workspace/zephyr_apps/smp_svr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/Users/0xff19/zephyr-sdk-0.14.2/arm-zephyr-eabi/arm-zephyr-eabi -imacros /Users/0xff19/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/Users/0xff19/Workspace/zephyr_apps/smp_svr=CMAKE_SOURCE_DIR -fmacro-prefix-map=/Users/0xff19/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/Users/0xff19/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /Users/0xff19/Workspace/zephyr_apps/smp_svr/src/main.c
In file included from /Users/0xff19/zephyrproject/zephyr/include/zephyr/arch/arm/aarch32/arch.h:20,
                 from /Users/0xff19/zephyrproject/zephyr/include/zephyr/arch/cpu.h:19,
                 from /Users/0xff19/zephyrproject/zephyr/include/zephyr/kernel_includes.h:33,
                 from /Users/0xff19/zephyrproject/zephyr/include/zephyr/kernel.h:17,
                 from /Users/0xff19/zephyrproject/zephyr/include/zephyr/zephyr.h:18,
                 from /Users/0xff19/Workspace/zephyr_apps/smp_svr/src/main.c:8:
/Users/0xff19/zephyrproject/zephyr/include/zephyr/devicetree/fixed-partitions.h:53:9: error: 'DT_COMPAT_fixed_partitions_LABEL_storage_PARTITION_ID' undeclared here (not in a function)
   53 |  DT_CAT(DT_COMPAT_fixed_partitions_LABEL_, label)
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/0xff19/zephyrproject/zephyr/include/zephyr/devicetree.h:3068:24: note: in definition of macro 'DT_CAT'
 3068 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/Users/0xff19/zephyrproject/zephyr/include/zephyr/storage/flash_map.h:264:2: note: in expansion of macro 'DT_FIXED_PARTITION_ID'
  264 |  DT_FIXED_PARTITION_ID(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
      |  ^~~~~~~~~~~~~~~~~~~~~
/Users/0xff19/zephyrproject/zephyr/include/zephyr/devicetree/fixed-partitions.h:53:2: note: in expansion of macro 'DT_CAT'
   53 |  DT_CAT(DT_COMPAT_fixed_partitions_LABEL_, label)
      |  ^~~~~~
/Users/0xff19/zephyrproject/zephyr/include/zephyr/storage/flash_map.h:264:24: note: in expansion of macro 'DT_NODE_BY_FIXED_PARTITION_LABEL'
  264 |  DT_FIXED_PARTITION_ID(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/0xff19/Workspace/zephyr_apps/smp_svr/src/main.c:58:25: note: in expansion of macro 'FLASH_AREA_ID'
   58 |  .storage_dev = (void *)FLASH_AREA_ID(storage),
      |                         ^~~~~~~~~~~~~
[61/266] Building C object zephyr/CMakeFiles/zephyr.dir/subsys/shell/shell.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /opt/homebrew/bin/cmake --build /Users/0xff19/Workspace/zephyr_apps/smp_svr/build
```

But if I choose `disco_l475_iot1` instead, like [this demo in OSSNA](https://youtu.be/bluz8uC7WEs?t=216): 

```
west build -p -b disco_l475_iot1 -- -DCONF_FILE="prj.conf overlay-bt.conf"
```

Then it compiles.
